Fix `black` dependency issues in CI

AnOctopus opened the pull request:
Pip does unreasonable things if you try to install packages in multiple
commands, and this is inherited by pipenv, so we put `black` in our dev
dependencies and only run the command that sets up both dev and test
dependencies in CI, rather than running dev and then combined.

AnOctopus committed:
Fix `black` dependency issues in CI

Pip does unreasonable things if you try to install packages in multiple
commands, and this is inherited by pipenv, so we put `black` in our dev
dependencies and only run the command that sets up both dev and test
dependencies in CI, rather than running dev and then combined.

AnOctopus committed:
Make pipenv-test-install verbose

AnOctopus committed:
Try removing sequential flag for pipenv so it resolves all at once

AnOctopus committed:
Directly specify torchvision bounds

tconkling commented on Makefile:
```diff
@@ -68,7 +68,7 @@ pipenv-test-install: lib/test-requirements.txt
 	cd lib ; \
 		cp Pipfile Pipfile.bkp ; \
-		pipenv install --dev --skip-lock --sequential -r test-requirements.txt ; \
+		pipenv install --dev --skip-lock --verbose -r test-requirements.txt ; \
```
The `--sequential` flag is important for "increasing determinism" in Pipenv (https://pipenv-fork.readthedocs.io/en/latest/advanced.html) - is there a good reason to drop it here?

(We added it while hunting down a dependency heisenbug in the distant past.)

AnOctopus replied:
I disabled it as an experiment, because the sequential flag, when combined with the underlying behavior of `pip install foo` `pip install bar` not being the same as `pip install foo bar`, means dependencies will almost never be resolved correctly.

tconkling replied:
This _sounds_ fragile, at least - `sequential` means "more deterministic", so removing it means "less deterministic". If this is the way forward, can we document in a short comment why it's so?

AnOctopus replied:
It is possible it won't be necessary to keep it, this PR is as much about finding how to make things work as it is to then get those changes merged in. But yes, if it seems necessary to keep it, I'll add a comment.

AnOctopus replied:
I think the documentation's claim that `sequential` makes it more deterministic is arguable, and I especially think that it achieves whatever determinism it does by very often doing the wrong thing

AnOctopus committed:
Specify typing-extensions version in pipenv install

Because otherwise pip installs a version of this that doesn't work with
anything else in the following step

AnOctopus committed:
Try removing skip-lock for pipenv install for CI

To see if this helps with ultimately resolving the wrong numpy version

AnOctopus committed:
Adjust TF test version bounds to fix grpcio version

AnOctopus committed:
Use non-broken pipenv version

AnOctopus committed:
Add back skip-lock to pipenv test install

Throwing things at the wall to see if it unbreaks pipenv in CI

AnOctopus committed:
What if we use a dedicated test pipfile

AnOctopus committed:
Actually include new test pipfile

AnOctopus committed:
Remove duplicate setuptools entry

AnOctopus committed:
Add sequential back to test pipfile command

AnOctopus commented:
With all the effort it took to do this, and the feeling that even though it works it is super fragile and painful to maintain, I don't want to merge something based on this. I'm going to try and do a smaller fix in another PR, and hopefully soon will try to have CI use `poetry` instead (which I use for my own development and have acceptable experiences with)
